larq_compute_engine/tflite/kernels/bconv2d.cc

```diff
@@ -209,43 +144,68 @@ TfLiteStatus Prepare(TfLiteContext* context, TfLiteNode* node) {
+    TF_LITE_ENSURE_EQ(context, SizeOfDimension(post_activation_multiplier, 0),
+                      conv_params->channels_out);
+    TF_LITE_ENSURE_EQ(context, SizeOfDimension(post_activation_bias, 0),
+                      conv_params->channels_out);
+  }
 
   if (output->type == kTfLiteInt8) {
+    // With 8-bit output, we allow int8 or float32 `post_activation` values
+    TF_LITE_ENSURE(context,
+                   post_activation_multiplier->type == kTfLiteFloat32 ||
+                       post_activation_multiplier->type == kTfLiteInt8);
+    TF_LITE_ENSURE(context, post_activation_bias->type == kTfLiteFloat32 ||
+                                post_activation_bias->type == kTfLiteInt8);
+
     TF_LITE_ENSURE_EQ(context, output->quantization.type,
                       kTfLiteAffineQuantization);
+
+    TF_LITE_ENSURE_MSG(
+        context,
+        conv_params->padding_type != kTfLitePaddingSame ||
+            conv_params->pad_value == 1,
+        "8-bit quantization is only supported with valid or one-padding");
   }
 
-  if (!write_bitpacked_output) {
-    conv_params->scaled_post_activation_multiplier.resize(
-        conv_params->channels_out);
-    conv_params->scaled_post_activation_bias.resize(conv_params->channels_out);
+  if (output->type == kTfLiteFloat32) {
+    // With float output, we require float `post_activation` values.
+    TF_LITE_ENSURE_EQ(context, post_activation_multiplier->type,
+                      kTfLiteFloat32);
+    TF_LITE_ENSURE_EQ(context, post_activation_bias->type, kTfLiteFloat32);
+  }
+
+  if (kernel_type == KernelType::kGenericRef) {
+    TF_LITE_ENSURE_MSG(
+        context,
+        conv_params->padding_type != kTfLitePaddingSame ||
+            conv_params->pad_value == 1,
+        "The reference kernel only supports valid or one-padding.");
   }
 
-  // Determine the output dimensions
+  // Determine the output dimensions and allocate the output buffer
   TfLiteIntArray* output_shape = TfLiteIntArrayCreate(4);
   output_shape->data[0] = SizeOfDimension(input, 0);
   output_shape->data[1] = out_height;
   output_shape->data[2] = out_width;
-  if (write_bitpacked_output) {
-    output_shape->data[3] =
-        ce::core::GetBitpackedSize(conv_params->channels_out);
-  } else {
-    output_shape->data[3] = conv_params->channels_out;
-  }
-
-  // Allocate the output buffer
-  TF_LITE_ENSURE_OK(context,
-                    context->ResizeTensor(context, output, output_shape));
-
-  if (kernel_type == KernelType::kGenericRef) {
-    // We only support one-padding or valid-padding in the reference
-    // implementation.
-    TF_LITE_ENSURE(context, !(conv_params->pad_value == 0 &&
-                              conv_params->padding_type == kTfLitePaddingSame));
-  }
+  output_shape->data[3] =
+      output->type == kTfLiteInt32
+          ? ce::core::GetBitpackedSize(conv_params->channels_out)
+          : output_shape->data[3] = conv_params->channels_out;
```

Comment:
I think this was a typo due to moving some lines around.
```suggestion
  output_shape->data[3] =
      output->type == kTfLiteInt32
          ? ce::core::GetBitpackedSize(conv_params->channels_out)
          : conv_params->channels_out;
```